frame/two-phase-election-provider/src/benchmarking.rs

```diff
@@ -0,0 +1,358 @@
+		T::Currency::make_free_balance_be(&receiver, initial_balance);
+		let ready: ReadySolution<T::AccountId> = Default::default();
+		let deposit: BalanceOf<T> = 10u32.into();
+		let reward: BalanceOf<T> = 20u32.into();
+
+		assert_ok!(T::Currency::reserve(&receiver, deposit));
+		assert_eq!(T::Currency::free_balance(&receiver), initial_balance - 10u32.into());
+	}: {
+		<TwoPhase<T>>::finalize_signed_phase_accept_solution(ready, &receiver, deposit, reward)
+	} verify {
+		assert_eq!(T::Currency::free_balance(&receiver), initial_balance + 20u32.into());
+		assert_eq!(T::Currency::reserved_balance(&receiver), 0u32.into());
+	}
+
+	finalize_signed_phase_reject_solution {
+		let receiver = account("receiver", 0, SEED);
+		let initial_balance = T::Currency::minimum_balance() * 10u32.into();
+		let deposit: BalanceOf<T> = 10u32.into();
+		T::Currency::make_free_balance_be(&receiver, initial_balance);
+		assert_ok!(T::Currency::reserve(&receiver, deposit));
+
+		assert_eq!(T::Currency::free_balance(&receiver), initial_balance - 10u32.into());
+		assert_eq!(T::Currency::reserved_balance(&receiver), 10u32.into());
+	}: {
+		<TwoPhase<T>>::finalize_signed_phase_reject_solution(&receiver, deposit)
+	} verify {
+		assert_eq!(T::Currency::free_balance(&receiver), initial_balance - 10u32.into());
+		assert_eq!(T::Currency::reserved_balance(&receiver), 0u32.into());
+	}
+
+	#[extra]
+	create_snapshot {
+		assert!(<TwoPhase<T>>::snapshot().is_none());
+	}: {
+		<TwoPhase::<T>>::create_snapshot()
+	} verify {
+		assert!(<TwoPhase<T>>::snapshot().is_some());
+	}
+
+	submit {
+		let c in 1 .. (T::MaxSignedSubmissions::get() - 1);
+
+		// the solution will be worse than all of them meaning the score need to be checked against all.
+		let solution = RawSolution { score: [(1000_0000u128 - 1).into(), 0, 0], ..Default::default() };
+
+		<CurrentPhase<T>>::put(Phase::Signed);
+		<Round<T>>::put(1);
+
+		for i in 0..c {
+			<SignedSubmissions<T>>::mutate(|queue| {
+				let solution = RawSolution { score: [(1000_0000 + i).into(), 0, 0], ..Default::default() };
+				let signed_submission = SignedSubmission { solution, ..Default::default() };
+				// note: this is quite tricky: we know that the queue will stay sorted here. The
+				// last will be best.
+				queue.push(signed_submission);
+			})
+		}
+
+		let caller = frame_benchmarking::whitelisted_caller();
+		T::Currency::make_free_balance_be(&caller,  T::Currency::minimum_balance() * 10u32.into());
+
+	}: _(RawOrigin::Signed(caller), solution, c)
+	verify {
+		assert!(<TwoPhase<T>>::signed_submissions().len() as u32 == c + 1);
+	}
+
+	submit_unsigned {
+		// number of votes in snapshot.
+		let v in 2000 .. 3000;
+		// number of targets in snapshot.
+		let t in 500 .. 800;
+		// number of assignments, i.e. compact.len(). This means the active nominators, thus must be
+		// a subset of `v` component.
+		let a in 500 .. 1500;
+		// number of desired targets. Must be a subset of `t` component.
+		let d in 200 .. 400;
```

Comment:
normally this would trigger to me that some more pallet config variables are needed to allow runtime developer to tell the benchmarks and pallets about their assumptions.

Could combine all of these into one struct or something, where the default impl are these numbers.

Reply:
Same below

Reply:
Good point, I totally agree. The current values are derived anecdotally by just one approach: If they are too small, they are inaccurate. I tried to make them large enough so that the results are somewhat stable. 

But I am still a bit suspicious if these weights are real + @thiolliere correctly pointed out that the weight difference of `submit_unsigned` and `feasibility_check` is a bit strange, as they are doing the same thing.

Reply:
Hmm as it turns out the lower bound of a range cannot be a varible and needs to be a literal? or am I mistaking? This works: `let v in 2000 .. T::BenchmarkingConfig::VOTERS[1];` while `let v in T::BenchmarkingConfig::VOTERS[1] .. T::BenchmarkingConfig::VOTERS[1];` doesn't.

Reply:
see: https://github.com/paritytech/substrate/pull/7902